IPIP/0298-denylists-for-ipfs-nodes-and-gateways.md

```diff
@@ -0,0 +1,242 @@
+
+A well-thought-out list schema can ease the implementation of denylists, and bring consensus on how content blocking works for gateways, and node operators in the community.
+
+## Architecture
+
+```mermaid
+sequenceDiagram
+  autonumber
+  actor User
+  participant IPFS Node
+  participant List Provider
+  note over IPFS Node: Discover List <CID> by List Provider
+  IPFS Node->>List Provider: ipfs://<CID>
+  List Provider->>IPFS Node: List
+  loop For each entry
+    IPFS Node->>IPFS Node: Add entry to rule engine
+  end
+  User->>IPFS Node: ipfs://bafy...badcid
+  IPFS Node->>IPFS Node: Process bafy...badcid through rule engine
+  IPFS Node->>User: 403 Forbidden
+```
+
+### List Provider
+
+A List Provider is an entity building and providing a list of content to be allowed/denied. They can provide their lists through the IPFS network or via [an HTTP Gateway endpoint](../http-gateways/README.md).
+
+### IPFS Node
+
+An IPFS node is a service that interacts with the IPFS network. This could be because it's providing content to the network, because it's retrieving content, or because it's serving IPFS hosted content through an HTTP interface.
+
+To interact with content lists, it operates a rule engine. The rule engine decides what action should be taken for a particular request.
+For instance, if CID bafy...badcid is denied, the bitswap engine should not request it, and access to bafy...badcid through a node exposed gateway are going to 403 forbidden.
+
+### User
+
+User is an actor that can make request to the IPFS Node. It's considered authenticated already.
+Access to the node can be made via an HTTP gateway, Bitswap, or another transport method exposed by the IPFS Node.
+
+It is understood that the User might be in control of the IPFS Node, and is able to configure the rule engine as it suits them.
+
+## Schemas
+
+### Definition
+
+The schemas below are defined as [IPLD schemas](https://ipld.io/specs/schemas/).
+
+```ipldsch
+type List struct {
+  Action Int # Suggested action to be taken
+  Categories optional [Int] # Content category entries should be associated to by default
+
+  Entries EntryChunk # Chain of Entries included in the list
+
+  Description optional String # Context to be shared with the list consumers
+}
+
+type EntryChunk struct {
+  Entries [Entry] # List of Entries objects
+  Next optional Link # CID for the next EntryChunk object. Entries referenced this way are not inlined, and possibly require an additional request for discovery.
```

Comment:
Having a single `Link` makes it:
- impossible to  compose a list from two preexisting ones (e.g., having "anti-phishing" meta-list composed of lists provided by gateways)
- expensive to maintain a big list (instead of having a DAG, and only recomputing changed branch, one needs to change all lists along the single chain).


Would it be ok to change this field into array of links?

```suggestion
  Next optional [Link] # List of CIDs for the next EntryChunk objects. Entries referenced this way are not inlined, and possibly require an additional request for discovery.
```

Reply:
What about having multiple `Entries [EntryChunk]` instead of multiple `Next [Link]`? I feel it makes it:
* Easier to read when lists are composed: it's simply a new EntryChunk.
* Does not modify a specific EntryChunk upon the addition of a new list.

Reply:
- Yes, `[EntryChunk]` also works.  
  - Remaining problem around `Link` is that there is no place to provide description for it– If i am aggregating third-party lists into a single meta-lists, would be nice to note why each link was included. Maybe add `Description optional String` to `EntryChunk` ?
- re: max size, i only meant max size per document, encouraging people to "shard"  at the same cut-off and making it easier for implementers to fail fast when consuming a single document. (total across lists is a separate problem).